README.md
# slippymath
[![Travis build status](https://travis-ci.org/MilesMcBain/slippymath.svg?branch=master)](https://travis-ci.org/MilesMcBain/slippymath) [![AppVeyor build status](https://ci.appveyor.com/api/projects/status/github/MilesMcBain/slippymath?branch=master&svg=true)](https://ci.appveyor.com/project/MilesMcBain/slippymath) [![lifecycle](https://img.shields.io/badge/lifecycle-maturing-blue.svg)](https://www.tidyverse.org/lifecycle/#maturing) [![CRAN status](https://www.r-pkg.org/badges/version/slippymath)](https://cran.r-project.org/package=slippymath)

R functions for dealing with slippy map tile servers (Google maps, Open Street
Map, Mapbox, Stamen et. al.)

The goal of this package is to take a bunch of logic that was lying around in
defunct packages and obscure wikis and make it #rstats cannon in a way that is
decoupled from any particular tile service or visualisation front end. Happy
mapping!

# Installation

```r
## Dev version
remotes::install_github("milesmcbain/slippymath")

## CRAN version
install.packages('slippymath')
```

# Usage


## Tile maths
Look at how many tiles you might need: 
```r
library(sf)

uluru_bbox <-
  st_bbox(c(xmin = 131.02084,
            xmax = 131.0535,
            ymin = -25.35461,
            ymax = -25.33568),
          crs = st_crs("+proj=longlat +ellps=WGS84"))

bbox_tile_query(uluru_bbox)

# A tibble: 17 x 8
#    x_min  y_min  x_max  y_max y_dim x_dim total_tiles  zoom
#    <dbl>  <dbl>  <dbl>  <dbl> <dbl> <dbl>       <dbl> <int>
# 1      3      2      3      2     1     1           1     2
# 2      6      4      6      4     1     1           1     3
# 3     13      9     13      9     1     1           1     4
# 4     27     18     27     18     1     1           1     5
# 5     55     36     55     36     1     1           1     6
# 6    110     73    110     73     1     1           1     7
# 7    221    146    221    146     1     1           1     8
# 8    442    293    442    293     1     1           1     9
# 9    884    586    884    586     1     1           1    10
#10   1769   1173   1769   1173     1     1           1    11
#11   3538   2346   3539   2346     1     2           2    12
#12   7077   4692   7078   4692     1     2           2    13
#13  14154   9384  14156   9385     2     3           6    14
#14  28309  18769  28312  18771     3     4          12    15
#15  56619  37538  56625  37542     5     7          35    16
#16 113239  75076 113251  75084     9    13         117    17
#17 226478 150153 226502 150168    16    25         400    18
```

Get a grid of slippy map tile coordinates for a bounding box, given a `zoom`, or
guess a zoom given a `max_tiles`:

```r
bbox_to_tile_grid(uluru_bbox, max_tiles = 15)

#$tiles
#       x     y
#1  28309 18769
#2  28310 18769
#3  28311 18769
#4  28312 18769
#5  28309 18770
#6  28310 18770
#7  28311 18770
#8  28312 18770
#9  28309 18771
#10 28310 18771
#11 28311 18771
#12 28312 18771
#
#$zoom
#[1] 15
#
#attr(,"class")
#[1] "tile_grid"
```

Fetch a grid using your favourite tile server API. Creep on many tile providers [here](http://leaflet-extras.github.io/leaflet-providers/preview/). 

Here's a Mapbox example:

```r

library(purrr)
library(curl)
library(glue)

tile_grid <- bbox_to_tile_grid(uluru_bbox, max_tiles = 15)

mapbox_query_string <-
  paste0("https://api.mapbox.com/v4/mapbox.satellite/{zoom}/{x}/{y}.jpg90",
         "?access_token=",
         Sys.getenv("MAPBOX_API_KEY"))

images <-
  pmap(tile_grid$tiles,
       function(x, y, zoom){
         outfile <- glue("{x}_{y}.jpg")
         curl_download(url = glue(mapbox_query_string),
                       destfile = outfile) 
         outfile 
       },
       zoom = tile_grid$zoom)

```

## Tile compositing

You can compose a list of images and a corresponding tile grid to a spatially
referenced raster, meaning you can plot over it with `tmap` etc.

```r
library(raster)
library(rgdal)

raster_out <- compose_tile_grid(tile_grid, images)

## A convenient wrapper for raster image exports using png::writePNG.
raster_to_png(raster_out, "uluru.png")

```
Result:
![](inst/media/uluru.png)

© <a href='https://www.mapbox.com/about/maps/'>Mapbox</a> © <a href='http://www.openstreetmap.org/copyright'>OpenStreetMap</a> <strong><a href='https://www.mapbox.com/map-feedback/' target='_blank'>Improve this map</a></strong>


NEWS.md
# slippymath 0.3.1

* Fixed a bug in `raster_to_png()` that corrupted the output in some cases.

# slippymath 0.3.0

Many function name changes in preparation for CRAN submisson. **These will certainly break existing `slippymath` code**. 

I have gone with a more explicit API since I feel users will mainly use
`slippymath` under the hood as a dependency. Therefore it is important they can
understand the API quickly rather than type it quickly - since they won't do
that very often.

Function names changing:
* `bb_tile_extent` -> `bbox_tile_extent`
* `bb_tile_query` -> `bbox_tile_query`
* `bb_to_tg` -> `bbox_to_tile_grid`
* `tg_composite` -> `compose_tile_grid`
* `tile_bb` -> `tile_bbox`
* `tg_bbs` -> `tile_grid_bboxes`


# slippymath 0.2.0

* Added a `NEWS.md` file to track changes to the package.
* BREAKING CHANGE. There were mixed lat/lon lon/lat conventions. I have settled on the lon/lat ordering. This means the following functions have changed name and argument order: 
    - `latlon_to_tilenum` -> `lonlat_to_tilenum`
    - `tilenum_to_latlon`  -> `tilenum_to_lonlat`



tests/testthat.R
library(testthat)
library(slippymath)

test_check("slippymath")


tests/testthat/test_normalise.R
context("test normalisation for PNG write")

test_that("raster normalisation works", {

  dummy_values <-
    rep(c(0, 64, 128, 255), 25)

  mat_5_20 <-
    raster::raster(nrows = 5,
                   ncols = 20,
                   vals = dummy_values)

  mat_20_5 <-
    raster::raster(nrows = 20,
                   ncols = 5,
                   vals = dummy_values)
  mat_10_10 <-
    raster::raster(nrows = 10,
                   ncols = 10,
                   vals = dummy_values)
  ## 3 layer
  expect_equal(
  raster::brick(mat_5_20,
                mat_5_20 * 0.1,
                mat_5_20 * 10) %>%
  normalise_raster() %>%
  range(),
  c(0, 1))

  ## 4 layer
  expect_equal(
    raster::brick(mat_5_20,
                  mat_5_20 * 0.1,
                  mat_5_20 * 10,
                  mat_5_20) %>%
    normalise_raster() %>%
    range(),
    c(0, 1))

  ## 2 layer
  expect_equal(
    raster::brick(mat_5_20,
                  mat_5_20 * 10) %>%
    normalise_raster() %>%
    range(),
    c(0, 1))

  ## 3 layer long
  expect_equal(
    raster::brick(mat_20_5,
                  mat_20_5 * 0.1,
                  mat_20_5 * 10) %>%
    normalise_raster() %>%
    range(),
    c(0, 1))

  ## 3 layer square
  expect_equal(
    raster::brick(mat_10_10,
                  mat_10_10 * 0.1,
                  mat_10_10 * 10) %>%
    normalise_raster() %>%
    range(),
    c(0, 1))

})


tests/testthat/test-tile_logic.R
context("test-tile_logic")


test_that("tile cascade works", {

  bbox <-  c(xmin = -180, xmax = 180, ymin = -85.0511, ymax = 85.0511)
  levs <- slippymath::bbox_tile_query(bbox, zoom_levels = 0:19)

  expect_equal(levs$zoom, 0:19)
  expect_equal(levs$total_tiles, (2^seq(0, 19))^2)
  expect_true(all(levs$x_min == 0))
  expect_true(all(levs$y_min == 0))
  expect_equal(levs$x_max,  2^seq(0, 19)-1)
  expect_equal(levs$x_max, levs$y_max)
  expect_equal(levs$x_max + 1, levs$x_dim)
  expect_equal(levs$x_dim, levs$y_dim)
})


test_that("tile cascade smaller extent works", {
  bbox <-  c(xmin = 140, xmax = 150,
                       ymin = -43, ymax = -41)
  levs_now <- slippymath::bbox_tile_query(bbox, zoom_levels = 0:19)
  #dput(levs)
  levs <- structure(list(x_min = c(0, 1, 3, 7, 14, 28, 56, 113, 227, 455,
                                   910, 1820, 3640, 7281, 14563, 29127, 58254, 116508, 233016, 466033),
                         y_min = c(0, 1, 2, 5, 10, 20, 40, 80, 160, 320, 640, 1280,
                                   2560, 5120, 10241, 20482, 40964, 81929, 163859, 327718),
                         x_max = c(0,1, 3, 7, 14, 29, 58, 117, 234, 469, 938, 1877, 3754, 7509, 15018,
                                  30037, 60074, 120149, 240298, 480597),
                         y_max = c(0, 1, 2, 5, 10, 20, 40, 80, 161, 323, 647, 1295, 2590, 5181, 10363, 20727,
                                   41454, 82909, 165819, 331638),
                         y_dim = c(1, 1, 1, 1, 1, 1, 1, 1, 2, 4, 8, 16, 31, 62, 123, 246, 491, 981, 1961, 3921),
                         x_dim = c(1, 1, 1, 1, 1, 2, 3, 5, 8, 15, 29, 58, 115, 229, 456, 911, 1821, 3642, 7283, 14565),
                         total_tiles = c(1, 1, 1, 1, 1, 2, 3, 5, 16, 60, 232, 928, 3565, 14198, 56088, 224106, 894111,
                                         3572802, 14281963, 57109365),
                         zoom = 0:19), row.names = c(NA, -20L), class = c("data.frame"))
 expect_equal(levs_now, levs)
})

test_that("whole zoom 0 tile logic works", {
  bbox <-  c(xmin = -180, xmax = 180, ymin = -85.0511, ymax = 85.0511)
  tg <- slippymath::bbox_to_tile_grid(bbox, zoom = 0)
  tg %>% expect_s3_class("tile_grid") %>% expect_named(c("tiles", "zoom"))
  expect_true(nrow(tg$tiles) == 1L)
  expect_true(tg$zoom == 0)
  })


R/cache.R
#' Downloader for tiles
#'
#' Tiles are cached with the native name of the source.
#'
#' `query_string` takes the form of a templae, see examples
#' @param x tiles object
#' @param query_string an api query template (see Details)
#' @param clobber if `TRUE` re download any existing tiles
#' @return WIP
#' @export
#'
#' @examples
#' mapbox_query_string <- paste0("https://api.mapbox.com/v4/mapbox.satellite/{zoom}/{x}/{y}.jpg90",
#' "?access_token=",
#' Sys.getenv("MAPBOX_API_KEY"))
down_loader <- function(x, query_string, clobber = FALSE) {
    purrr::pmap(x$tiles,
         function(x, y, zoom){
           api_query <- glue::glue(query_string)
           outfile <- url_to_cache(api_query)
           if (!file.exists(outfile) || clobber) {
             cachedir <- fs::path_dir(outfile)

             if (!fs::dir_exists(cachedir)) fs::dir_create(cachedir, recursive = TRUE)

             ## FIXME: need to error on no API_KEY present
             curl::curl_download(url = api_query,
                         destfile = outfile)
           }
           outfile
         },
         zoom = x$zoom)
}

slippy_cache <- function() {
  cache <- file.path(rappdirs::user_cache_dir(), ".slippymath1")
  if (!fs::dir_exists(cache)) fs::dir_create(cache)
  cache
}

url_to_cache <- function(x) {
  base_filepath <- file.path(slippy_cache(), gsub("^//", "", gsub("^https\\:", "", gsub("^https\\:", "", x))))
  ## chuck off any ? junk
  unlist(lapply(strsplit(base_filepath, "\\?"), "[", 1L))
}


R/utils.R
radians <- function(angle_deg) angle_deg * pi / 180

degrees <- function(angle_rad) (angle_rad * 180) / pi

is_bbox <- function(obj){
  inherits(obj, "bbox") ||
    is_bbox_vector(obj)
}

is_bbox_vector <- function(obj){
  is.vector(obj) &&
    length(obj) == 4 &&
    length(setdiff(names(obj), c("xmin", "xmax", "ymin", "ymax"))) == 0
}

is_tile_grid <- function(obj) inherits(obj, "tile_grid")

assert_bbox <- function(obj){
  if (!is_bbox(obj)){
    stop("bbox needs to be created with st_bbox() or be a named numeric vector with names: xmin, xmax, ymin, and ymax")
  } 
}
sp_bbox_to_sf <- function(sp_bbox) {
  stats::setNames(as.vector(sp_bbox), c("xmin", "ymin", "xmax", "ymax"))
}

##' Write a raster to PNG
##'
##' This function is a convenience wrapper for
##' writing rasters in PNG format.
##'
##' @title raster_to_png
##' @param tile_raster the raster to write to PNG
##' @param file_path the path to write the raster
##' @return nothing.
##' @export
raster_to_png <- function(tile_raster, file_path){

  if (!inherits(tile_raster, "RasterBrick")){
    stop("tile raster must be a RasterBrick. This is output from tg_composite().") 
  }

  if (!(raster::nlayers(tile_raster) %in% c(1,3,4))){
    stop("tile raster in unrecognised format - nlayers is not 1,3, or 4).")
  }

  ## png expects 0-1 values, so normalise:
  normalised_raster <-
    normalise_raster(tile_raster)

  png::writePNG(normalised_raster,
                target = file_path)
}


normalise_raster <- function(a_raster){
  sweep(raster::as.array(a_raster),
        MARGIN = 3,
        STATS = a_raster@data@max,
        FUN = "/")
}

lol_to_df <- function(lol){
  lov <- purrr::map(purrr::transpose(lol), unlist)
  purrr::lift_dl(data.frame)(lov)
}


R/mercator.R
globalVariables(c("A", "MAXEXTENT"), "slippymath") #ignore this in R CMD checks

# Convert lon/lat values to 900913 x/y.
A <- 6378137
MAXEXTENT <- 20037508.342789244;
# clamp values within range min/max
sm_clamp <- function(x, mn, mx) {
  x[x < mn] <- mn
  x[x > mx] <- mx
  x
}

# https://github.com/mapbox/sphericalmercator/blob/3f61128523aed26904be33463af87a142752a4ad/sphericalmercator.js#L175
#' Transform between spherical Mercator and longitude/latitude
#'
#' @param ll matrix of longitude / latitude
#' @param xy matrix of x / y Mercator
#' @return matrix of coordinates transformed forward or inverse
#' @examples
#' uluru_lonlat <- matrix(c(131.0325162,
#'                          -25.3448562),
#'                        nrow = 1)
#'
#' lonlat_to_merc(uluru_lonlat)
#'
#' uluru_merc <- matrix(c(14586472.958481,
#'                        -2918162.223463),
#'                      nrow = 1)
#'
#' merc_to_lonlat(uluru_merc)
#' @export
#' @aliases merc_to_lonlat
lonlat_to_merc <- function(ll) {
  xy <-
      cbind(A * radians(ll[,1]),
          A * log(tan((pi*0.25) + (0.5 * radians(ll[,2])))))
  ## if xy value is beyond maxextent (e.g. poles), return maxextent.
  xy[, 1] <- sm_clamp(xy[, 1], -MAXEXTENT, MAXEXTENT)
  xy[, 2] <- sm_clamp(xy[, 2], -MAXEXTENT, MAXEXTENT)
  xy
}

## Convert 900913 x/y values to lon/lat.
#' @name lonlat_to_merc
#' @export
merc_to_lonlat <- function(xy) {
  xy <- cbind(degrees(xy[, 1] / A),
              degrees(((pi*0.5) - 2.0 * atan(exp(-xy[, 2] / A)))))
  xy
}

##' Are points in meters within Mercator extent?
##'
##' When doing maths with Mercator coordinates in m, you can end up outside the
##' Mercator extent with an undefined coordinate. This function returns true if
##' all xy lie within the Mercator extent.
##' @title within_mercator_extent
##' @param xy a matrix of Mercator xy coordinates.
##' @return TRUE or FALSE
##' @examples
##' stray <- matrix(c(20037509,
##'                    -2918162.223463),
##'                  nrow = 1)
##'
##' within_merc_extent(stray)
##' @export
within_merc_extent <- function(xy){
  all(xy <= MAXEXTENT) & all(xy >= -MAXEXTENT)
}

##' Truncate coordinate to Mercator extent.
##'
##' If a point in m lies outside the Mercator extent, this function can be used
##' to truncate it to the boundary of the extent.
##'
##' @title merc_truncate
##' @param xy a matrix of Mercator XY points.
##' @return a matrix of XY points.
##' @examples
##' stray <- matrix(c(20037509,
##'                    -2918162.223463),
##'                  nrow = 1)
##'
##' merc_truncate(stray)
##'
##' @export
merc_truncate <- function(xy){
  xy[, 1] <- sm_clamp(xy[, 1], -MAXEXTENT, MAXEXTENT)
  xy[, 2] <- sm_clamp(xy[, 2], -MAXEXTENT, MAXEXTENT)
  xy
}


R/georeference.R
#' @importFrom tibble tibble
#' @importFrom dplyr filter
spherical_mercator <- function(provider) {
  #MAXEXTENT is the bounds between [-180, 180] and [-85.0511, 85.0511]
  tibble::tibble(provider = "mapbox",
                 MAXEXTENT = 20037508.342789244,
                 A = 6378137.0, B = 6378137.0,
                 crs = glue::glue("+proj=merc +a={A} +b={A}")) %>%
    dplyr::filter(provider == provider)
}
#' @importFrom stats setNames
mercator_tile_extent <- function(tile_x, tile_y, zoom, tile_size = 256) {
  if (any(!c(length(tile_x), length(tile_y), length(zoom), length(tile_size)) == 1)) {
    stop("tile_x, tile_y, zoom, tile_size must all be of length 1")
  }
  params <- spherical_mercator(provider = "mapbox")
  params <- params[1, ]  ## FIXME: param query should provide a unique set, this is WIP
  #st_transform(st_as_sfc(my_bbox), glue("+proj=merc +a={A} +b={A}"))
  MAXEXTENT <- params$MAXEXTENT
  A <- params$A
  ## literal width/height of a square tile at zoom = 0
  z0_size <- (MAXEXTENT * 2)/tile_size
  xlim <- -MAXEXTENT + (tile_x + c(0, 1)) * (z0_size/(2^zoom)) * tile_size
  ## upside down Ms. Jane
  ylim <- range(MAXEXTENT - (tile_y + c(0, 1) - 0) * (z0_size/(2^zoom)) * tile_size)
  stats::setNames(c(xlim, ylim), c("xmin", "xmax", "ymin", "ymax"))
}


R/slippymath.R
globalVariables(c(".global_sm_env"), "slippymath") #ignore this in R CMD checks
.global_sm_env <- new.env(parent=emptyenv())

.global_sm_env$WEB_MERCATOR_CRS <- structure(list(epsg = 3857L, proj4string = "+proj=merc +a=6378137 +b=6378137 +lat_ts=0.0 +lon_0=0.0 +x_0=0.0 +y_0=0 +k=1.0 +units=m +nadgrids=@null +wktext +no_defs"), class = "crs")
## sf::st_crs(3857)

.global_sm_env$LONLAT_CRS <- structure(list(epsg = 4326L, proj4string = "+proj=longlat +datum=WGS84 +no_defs"), class = "crs")
## sf::st_crs(4326)

##' Convert longitude and latitude to slippy tile numbers
##'
##' Returns the Open Street Map slippy map tile numbers (x, y) the
##' supplied latitude and longitude fall on, for a given zoom level.
##'
##' The point specified by lon_deg` and `lat_deg` is assumed to be in ESPG:4326
##' coordinate reference system.
##'
##' @title lonlat_to_tilenum
##' @param lon_deg degrees longitude for point
##' @param lat_deg degrees latitude for point
##' @param zoom zoom level for tile calculation. Increasing zoom increases the
##'   number of tiles.
##' @return a list containing `x` and `y` - the tile numbers.
##' @examples
##' lonlat_to_tilenum(
##'  lon = 13.37771496361961,
##'  lat = 52.51628011262304,
##'  zoom = 17
##' )
##' @export
lonlat_to_tilenum <- function(lon_deg, lat_deg, zoom){
    ## Implementing slippy map spec as per https://wiki.openstreetmap.org/wiki/Slippy_map_tilenames
    lon_rad <- radians(lon_deg)
    lat_rad <- radians(lat_deg)

    x <- lon_rad
    y <- asinh(tan(lat_rad))

    x <- (1 + (x / pi))/2
    y <- (1 - (y / pi))/2

    n_tiles <- 2^zoom

    ## The values are clamped to prevent problems at the extent boundaries. Eg 180
    ## degrees lon which would lon_rad of pi.
    xtile <- sm_clamp(floor(x * n_tiles), 0, n_tiles-1)
    ytile <- sm_clamp(floor(y * n_tiles), 0, n_tiles-1)

    list(x = xtile, y = ytile)
}

##' Convert slippy map tiles numbers to latitude and longitude
##'
##' Returns the latitude and longitude of the top left corner of a slippy map tile
##' specified by `x`, `y` for a given zoom level.
##' @title  tilenum_to_lonlat
##' @param x slippy map tile number in x domain (left to right)
##' @param y slippy map tile number in y domain (top to bottom)
##' @param zoom the zoom level for the calculation. Increasing zoom increases
##'   the number of tiles.
##' @return a list containing `lat` and `lon` - latitude and longitude.
##' @examples
##' tilenum_to_lonlat(
##'  x = 70406,
##'  y = 42987,
##'  zoom = 17
##' )
##' @export
tilenum_to_lonlat <- function(x, y, zoom){
    n_tiles <- 2^zoom

    lon_rad <- (((x / n_tiles) * 2) - 1) * pi

    merc_lat <- (1 - ((y / n_tiles) * 2)) * pi
    lat_rad <- atan(sinh(merc_lat))

    list(lon = degrees(lon_rad),
         lat = degrees(lat_rad))
}

##' Bounding box to tile grid
##'
##' Calculate a slippy map tile grid that will fit a supplied bounding box.
##'
##' The grid is returned as part of a tile_grid object that contains a
##' data.frame of x,y tile numbers and zoom level.
##'
##' The tile grid can be calculated for a given zoom level or for the deepest
##' zoom that ensures the number of tiles is less than or equal to `max_tiles`.
##'
##' If `zoom` and `max_tiles` are supplied together, then the max is still
##' enforced and the function will fail if more tiles are required for the given
##' zoom.
##'
##' @title bbox_to_tile_grid
##' @param bbox the bounding box to fit onto a grid of tiles. Must be either a
##'   'bbox' object created with sf::st_bbox or a vector of length 4 with names:
##'   'xmin', 'xmax', 'ymin', 'ymax'.
##' @param zoom Optional. The desired zoom level.
##' @param max_tiles Optional. The maximum number of tiles the grid may occupy.
##' @return a 'tile_grid' object containing 'tiles' and 'zoom'
##' @examples
##' tibrogargan<- c(xmin = 152.938485, ymin = -26.93345, xmax = 152.956467, 
##'                ymax = -26.921463)
##'
##' ## Get a grid of the minimum number of tiles for a given zoom.
##' bbox_to_tile_grid(tibrogargan, zoom = 15)
##'
##' ## get a grid of at most 12 tiles, choosing the most detailed zoom possible.
##' bbox_to_tile_grid(tibrogargan, max_tiles = 12)
##' @export
bbox_to_tile_grid <- function(bbox,
                     zoom = NULL,
                     max_tiles = NULL){

    if (purrr::is_null(zoom) && purrr::is_null(max_tiles)){
        stop("at least one of the zoom or max_tiles arugments must be supplied")
    }

    ## No zoom, we'll do a query and choose the best zoom for the max_tiles budget
    if (purrr::is_null(zoom)){
        tile_query <- bbox_tile_query(bbox, zoom_levels = 0:19)
        suitable_zooms <- tile_query$total_tiles <= max_tiles
        zoom <- tile_query$zoom[max(which(suitable_zooms))]
    }

    tile_extent <- bbox_tile_extent(bbox, zoom)

    x_tiles <- tile_extent$x_min:tile_extent$x_max
    y_tiles <- tile_extent$y_min:tile_extent$y_max

    if(!purrr::is_null(max_tiles) && (length(x_tiles) * length(y_tiles)) > max_tiles){
        stop("Bounding box needed more than max_tiles at specified zoom level. Check with bbox_tile_query(bbox)")
    }

    tile_grid <-
        list(
            tiles = expand.grid(x = x_tiles, y = y_tiles),
            zoom = zoom)
    class(tile_grid) <- "tile_grid"

    tile_grid
}

##' Bounding box tile query
##'
##' Determines how many tiles the bounding box would occupy for a range of zooms. Useful for working out what is a reasonable zoom to work at. Each tile is a separate request from the server.
##'
##' Tiles are typically 256x256 pixels and are tens of Kb in size, you can get some sense of the data from the query also.
##'
##' @title bbox_tile_query
##' @param bbox a bbox object created by `sf::st_bbox`, or a vector with names
##'   'xmin', 'xmax', 'ymin', 'ymax'
##' @param zoom_levels a numeric vector of zoom levels to calculate tile usage for.
##' @return a data frame containing tile usage information for the bounding box
##'   at each zoom level.
##' @examples
##' tibrogargan<- c(xmin = 152.938485, ymin = -26.93345, xmax = 152.956467, 
##'                ymax = -26.921463)
##'
##' bbox_tile_query(tibrogargan)
##' @export
bbox_tile_query <- function(bbox, zoom_levels = 2:18){

    extents_at_zooms <- purrr::map(zoom_levels,
                                   ~bbox_tile_extent(bbox, .))

    extents_at_zooms <- lol_to_df(extents_at_zooms)

    extents_at_zooms$y_dim <-
        abs(extents_at_zooms$y_max - extents_at_zooms$y_min) + 1
    extents_at_zooms$x_dim <-
        abs(extents_at_zooms$x_max - extents_at_zooms$x_min) + 1
    extents_at_zooms$total_tiles <-
        extents_at_zooms$y_dim * extents_at_zooms$x_dim
    extents_at_zooms$zoom <- zoom_levels

    extents_at_zooms
}

##' Convert a bounding box from latitude and longitude to tile numbers
##'
##' This function creates an analog of a bounding box but in tile numbers. It
##' returns the min and max x and y tile numbers for a tile grid that would fit
##' the bounding box for a given zoom level.
##'
##' @title bbox_tile_extent
##' @param bbox a bbox object created by `sf::st_bbox`, or a vector with names
##'   'xmin', 'xmax', 'ymin', 'ymax'
##' @param zoom zoom level to calculate the tile grid on.
##' @return a list of `x_min`, `y_min`, `x_max`, `y_max`
##' @examples
##' tibrogargan<- c(xmin = 152.938485, ymin = -26.93345, xmax = 152.956467, 
##'                ymax = -26.921463)
##' bbox_tile_extent(tibrogargan, zoom = 15)
##' @export
bbox_tile_extent <- function(bbox, zoom){
    assert_bbox(bbox)

    min_tile <- lonlat_to_tilenum(lat_deg = bbox["ymin"],
                                  lon_deg = bbox["xmin"], zoom)
    max_tile <- lonlat_to_tilenum(lat_deg = bbox["ymax"],
                                  lon_deg = bbox["xmax"], zoom)

    list(x_min = min_tile$x,
         y_min = max_tile$y,
         x_max = max_tile$x,
         y_max = min_tile$y)

    ## Note tile numbers start at 0 in the north and increase going south. The
    ## have the opposite polarity to latitude which increases going north. This is
    ## why y_min = max_tile$y here.
}

##' Calculate the bounding box for a tile in latitude and longitude
##'
##' Given a slippy maps tile specified by `x`, `y`, and `zoom`, return the
##' an `sf` bounding box object for the tile with units in metres using the
##' EPSG:3857 coordinate reference system (Web Mercator).
##'
##' @title tile_bbox
##' @param x slippy map tile x number
##' @param y slippy map tile y number
##' @param zoom zoom level for tile
##' @return an sf bbox object.
##' @examples
##' ## return an sf style bbox object in with epsg and proj4string
##' tile_bbox(x = 30304, y = 18929, zoom = 15)
##' @export
tile_bbox <- function(x, y, zoom){

    bottom_left <-
        lonlat_to_merc(t(as.matrix(unlist(tilenum_to_lonlat(x, y+1, zoom)))))
    
    top_right <-
        lonlat_to_merc(t(as.matrix(unlist(tilenum_to_lonlat(x+1, y, zoom)))))

    structure(c(xmin = bottom_left[[1]],
                ymin = bottom_left[[2]],
                xmax = top_right[[1]],
                ymax = top_right[[2]]),
              class = "bbox",
              crs = .global_sm_env$WEB_MERCATOR_CRS)
 }   




##' Get tile grid bounding boxes
##'
##' Given an tile_grid object like that returned from `bbox_to_tile_grid`, return
##' a list
##' of sf style bounding box objects, one for each tile in the grid, in the same order
##' as tiles in `tile_grid$tiles`.
##'
##' The bounding box units are metres in the EPSG:3857 coordinate reference
##' system (Web Mercator).
##'
##' @title tile_grid_bboxes
##' @param tile_grid a tile_grid object, likely returned from `bbox_to_tile_grid`
##' @return a list of sf bounding box objects in the corresponding order to the
##'   tiles in `tile_grid`
##' @examples
##' 
##' tibrogargan<- c(xmin = 152.938485, ymin = -26.93345, xmax = 152.956467, 
##'                ymax = -26.921463)
##' 
##' tibrogargan_grid <- bbox_to_tile_grid(tibrogargan, zoom = 15)
##'
##' tile_grid_bboxes(tibrogargan_grid)
##' @export
tile_grid_bboxes <- function(tile_grid){
    if(!is_tile_grid(tile_grid)) stop("tile_grid must be of class tile_grid - output from bbox_to_tile_grid()")

    purrr::pmap(.l = tile_grid$tiles,
                .f = tile_bbox,
                zoom = tile_grid$zoom)
}

##' Compose a list of images using tile_grid data.
##'
##' Given a tile_grid object and a list of images, compose the images into a
##' single spatially referenced RasterBrick object.
##'
##' The list of images is assumed to be in corresponding order to the tiles in
##' the tile_grid object.
##'
##' The returned object uses the Web Mercator projection, EPSG:3857, which is
##' the native crs of the tiles.
##'
##' @title compose_tile_grid
##' @param tile_grid a tile_grid object, likely returned from `bbox_to_tile_grid`
##' @param images a list of character strings defining paths to images. Matched to tiles in tile_grid based on list position.
##' @return a spatially referenced raster.
##' @export
compose_tile_grid <- function(tile_grid, images){

    bricks <-
        purrr::pmap(.l = list(x = tile_grid$tiles$x,
                              y = tile_grid$tiles$y,
                              image = images),
                    .f = function(x, y, image, zoom){
                        bbox <- tile_bbox(x, y, zoom)
                        raster_img <-
                            raster::brick(image,
                                          crs = attr(bbox, "crs")$proj4string)
                        raster::extent(raster_img) <-
                            raster::extent(bbox[c("xmin", "xmax", "ymin", "ymax")])
                        raster_img
                    },
                    zoom = tile_grid$zoom)

    geo_refd_raster <- do.call(raster::merge, bricks)
    
    geo_refd_raster
}
